modules/reindex/src/test/java/org/elasticsearch/index/reindex/DeleteByQueryConcurrentTests.java

```diff
@@ -0,0 +1,143 @@
+import org.elasticsearch.index.query.MatchQueryBuilder;
+import org.elasticsearch.index.query.QueryBuilders;
+
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static org.elasticsearch.index.query.QueryBuilders.matchQuery;
+import static org.elasticsearch.index.query.QueryBuilders.termQuery;
+import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertAcked;
+import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertHitCount;
+import static org.hamcrest.Matchers.equalTo;
+
+public class DeleteByQueryConcurrentTests extends ReindexTestCase {
+
+    public void testConcurrentDeleteByQueriesOnDifferentDocs() throws Throwable {
+        createIndex("test");
+        ensureGreen();
+
+        final Thread[] threads =  new Thread[scaledRandomIntBetween(2, 5)];
+        final long docs = randomIntBetween(1, 50);
+        for (int i = 0; i < docs; i++) {
+            for (int j = 0; j < threads.length; j++) {
+                index("test", "test", String.valueOf(i * 10 + j), "field", j);
+            }
+        }
+        refresh();
+        assertHitCount(client().prepareSearch("test").setSize(0).get(), docs * threads.length);
+
+        final CountDownLatch start = new CountDownLatch(1);
+        final AtomicReference<Throwable> exceptionHolder = new AtomicReference<>();
+
+        for (int i = 0; i < threads.length; i++) {
+            final int threadNum = i;
+            assertHitCount(client().prepareSearch("test").setSize(0).setQuery(QueryBuilders.termQuery("field", threadNum)).get(), docs);
+
+            Runnable r = () -> {
+                try {
+                    start.await();
+
+                    assertThat(deleteByQuery().source("_all").filter(termQuery("field", threadNum)).refresh(true).get(),
+                            matcher().deleted(docs));
+                } catch (InterruptedException e) {
+                    Thread.currentThread().interrupt();
+                } catch (Throwable e) {
+                    exceptionHolder.set(e);
+                    Thread.currentThread().interrupt();
+                }
+            };
+            threads[i] = new Thread(r);
+            threads[i].start();
+        }
+
+        start.countDown();
+        for (Thread thread : threads) {
+            thread.join();
+        }
+
+        Throwable assertionError = exceptionHolder.get();
+        if (assertionError != null) {
+            throw assertionError;
+        }
+
+        for (int i = 0; i < threads.length; i++) {
+            assertHitCount(client().prepareSearch("test").setSize(0).setQuery(QueryBuilders.termQuery("field", i)).get(), 0);
+        }
+    }
+
+    public void testConcurrentDeleteByQueriesOnSameDocs() throws Throwable {
+        assertAcked(prepareCreate("test").setSettings(Settings.builder().put("index.refresh_interval", -1)));
```

Comment:
I've learned that any test that relies on refresh interval is going to fail. Some item something is going to force a refresh on us when we least expect it. Is this test ok with that?